cluster-autoscaler/core/static_autoscaler_test.go

```diff
@@ -1120,6 +1121,129 @@ func TestStaticAutoscalerRunOnceWithFilteringOnUpcomingNodesEnabledNoScaleUp(t *
+			deleteFinished <- true
+			return ret
+		},
+		nil, nil, nil, nil)
+	provider.AddNodeGroup("ng1", 1, 10, 1)
+	provider.AddNode("ng1", n1)
+	ng1 := reflect.ValueOf(provider.GetNodeGroup("ng1")).Interface().(*testprovider.TestNodeGroup)
+	assert.NotNil(t, ng1)
+	assert.NotNil(t, provider)
+
+	// Create context with mocked lister registry.
+	options := config.AutoscalingOptions{
+		NodeGroupDefaults: config.NodeGroupAutoscalingOptions{
+			ScaleDownUnneededTime:         time.Minute,
+			ScaleDownUnreadyTime:          time.Minute,
+			ScaleDownUtilizationThreshold: 0.5,
+			MaxNodeProvisionTime:          10 * time.Second,
+		},
+		EstimatorName:           estimator.BinpackingEstimatorName,
+		EnforceNodeGroupMinSize: true,
+		ScaleDownEnabled:        true,
+		MaxNodesTotal:           1,
+		MaxCoresTotal:           10,
+		MaxMemoryTotal:          100000,
+	}
+	processorCallbacks := newStaticAutoscalerProcessorCallbacks()
+
+	clientset := fake.NewSimpleClientset(n1, n2)
+	context, err := NewScaleTestAutoscalingContext(options, clientset, nil, provider, processorCallbacks, nil)
+	assert.NoError(t, err)
+
+	setUpScaleDownActuator(&context, options)
+
+	listerRegistry := kube_util.NewListerRegistry(allNodeLister, readyNodeLister, allPodListerMock, podDisruptionBudgetListerMock, daemonSetListerMock,
+		nil, nil, nil, nil)
+	context.ListerRegistry = listerRegistry
+
+	clusterStateConfig := clusterstate.ClusterStateRegistryConfig{
+		OkTotalUnreadyCount: 1,
+	}
+	processors := NewTestProcessors(&context)
+	clusterState := clusterstate.NewClusterStateRegistry(provider, clusterStateConfig, context.LogRecorder, NewBackoff(), nodegroupconfig.NewDefaultNodeGroupConfigProcessor(options.NodeGroupDefaults))
+	sdPlanner, sdActuator := newScaleDownPlannerAndActuator(&context, processors, clusterState)
+	suOrchestrator := orchestrator.New()
+	suOrchestrator.Initialize(&context, processors, clusterState, taints.TaintConfig{})
+
+	context.AutoscalingOptions.MaxBulkSoftTaintCount = 10
+	context.AutoscalingOptions.MaxBulkSoftTaintTime = 3 * time.Second
+
+	autoscaler := &StaticAutoscaler{
+		AutoscalingContext:    &context,
+		clusterStateRegistry:  clusterState,
+		lastScaleUpTime:       time.Now(),
+		lastScaleDownFailTime: time.Now(),
+		scaleDownPlanner:      sdPlanner,
+		scaleDownActuator:     sdActuator,
+		scaleUpOrchestrator:   suOrchestrator,
+		processors:            processors,
+		processorCallbacks:    processorCallbacks,
+		initialized:           true,
+	}
+
+	// Node from selected node groups can get their deletion candidate taints removed.
+	readyNodeLister.SetNodes([]*apiv1.Node{n1})
+	allNodeLister.SetNodes([]*apiv1.Node{n1})
+	allPodListerMock.On("List").Return([]*apiv1.Pod{p1}, nil).Twice()
+	daemonSetListerMock.On("List", labels.Everything()).Return([]*appsv1.DaemonSet{}, nil)
+	podDisruptionBudgetListerMock.On("List").Return([]*policyv1.PodDisruptionBudget{}, nil)
+
+	err = autoscaler.RunOnce(time.Now().Add(5 * time.Hour))
```

Comment:
Can you make this test table-based? Also - optional - there's a ton of dependencies here, which suggests maybe it would be cleaner to migrate to a new file that doesn't require most of the setup above and can be tested more easily.

Reply:
Yeah I already tried to remove a bunch of those dependencies but given I don't have much knowledge about all the interactions, I wasn't sure about exactly what needed to be removed.
Regarding the table test part, I can do that but that will mean there are 2 different testing styles in the codebase, so I'm not sure if you're OK with that? Maybe you're already planning on moving to table tests?

Reply:
Yeah, moving everything to table based is a lot of effort, so for now we're just encouraging this approach in new or modified tests.

Reply:
And by getting rid of dependencies I meant factoring out this code to a dedicated package/object. You don't really need to understand all of the deps of static_autoscaler - simply depend only on what your new code needs to run.

Reply:
Right, I've made the test table-based and tried removing as many deps as possible. Unfortunately I can't use the setupAutoscaler as it does not allow the caller to set a specific ClientSet, which I need in this test.